### Limitation(s) of sklearn’s non-negative matrix factorization library. 
In this part I'll upload the movie rating data and use NMF from sklearn in order to make rating predictions on the test set to see the RMSE.

## Part 1
Load the movie ratings data (as in the HW3-recommender-system) and use matrix factorization technique(s) and predict the missing ratings from the test data. Measure the RMSE. You should use sklearn library.

In [30]:
import pandas as pd
import numpy as np

In [31]:
MV_users = pd.read_csv('data/users.csv')
MV_movies = pd.read_csv('data/movies.csv')
train = pd.read_csv('data/train.csv')
test = pd.read_csv('data/test.csv')

Describe the data to remeber how it's composed.

In [10]:
print('MV_users data:')
print(MV_users.head())
print('-'*100)
print('MV_movies data:')
print(MV_movies.head())
print('-'*100)
print('train data:')
print(train.head())
print('-'*100)
print('test data:')
print(test.head())
print('-'*100)

MV_users data:
   uID gender  age  accupation    zip
0    1      F    1          10  48067
1    2      M   56          16  70072
2    3      M   25          15  55117
3    4      M   45           7  02460
4    5      M   25          20  55455
----------------------------------------------------------------------------------------------------
MV_movies data:
   mID                        title  year  Doc  Com  Hor  Adv  Wes  Dra  Ani  \
0    1                    Toy Story  1995    0    1    0    0    0    0    1   
1    2                      Jumanji  1995    0    0    0    1    0    0    0   
2    3             Grumpier Old Men  1995    0    1    0    0    0    0    0   
3    4            Waiting to Exhale  1995    0    1    0    0    0    1    0   
4    5  Father of the Bride Part II  1995    0    1    0    0    0    0    0   

   ...  Chi  Cri  Thr  Sci  Mys  Rom  Fil  Fan  Act  Mus  
0  ...    1    0    0    0    0    0    0    0    0    0  
1  ...    1    0    0    0    0    0    0

Now, I'll use the NMF to predict the ratings in the test set.

In [32]:
from sklearn.decomposition import NMF
from sklearn.metrics import mean_squared_error

First, I'm building the user x movies rating matrix, this is the one for applying the NMF.

In [64]:
user_movies_matrix = train.pivot(index='uID', columns='mID', values='rating').fillna(0)

Notice, that I'm filling the non rated movies with '3'.

In [65]:
user_movies_matrix

mID   1     2     3     4     5     6     7     8     9     10    ...  3943  \
uID                                                               ...         
1      5.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
5      0.0   0.0   0.0   0.0   0.0   2.0   0.0   0.0   0.0   0.0  ...   0.0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
6036   0.0   0.0   0.0   2.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
6037   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
6038   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
6039   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
6040   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

mID   3944  3945  3946  3947  3948  3949  3950  3951  3952  
uID                                                         
1      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
6036   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
6037   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
6038   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
6039   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
6040   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[6040 rows x 3664 columns]

As we can see, the size is 6040 x 3664, that's because there are 6040 users and 3664 unique movies in the train set, let's see.

In [66]:
print('Unique users in train:')
print(len(train['uID'].unique()))
print('Unique movies in train:')
print(len(train['mID'].unique()))

Unique users in train:
6040
Unique movies in train:
3664


In [67]:
nmf = NMF(n_components=20)
user_matrix = nmf.fit_transform(user_movies_matrix)
movies_matrix = nmf.components_

/Users/salvadorbarcenas/opt/anaconda3/lib/python3.9/site-packages/sklearn/decomposition/_nmf.py:1710: ConvergenceWarning: Maximum number of iterations 200 reached. Increase it to improve convergence.
  warnings.warn(


In [68]:
user_matrix.shape

(6040, 20)

In [69]:
movies_matrix.shape

(20, 3664)

Now, let's build the matriz rating predictions based on the user_matrix and movies_matrix

In [70]:
R_pred = np.dot(user_matrix, movies_matrix)
R_pred = pd.DataFrame(R_pred, index=user_movies_matrix.index, columns=user_movies_matrix.columns)

In [71]:
R_pred

mID       1         2         3         4         5         6         7     \
uID                                                                          
1     1.775903  0.560754  0.007710  0.017489  0.011811  0.026577  0.012459   
2     1.487041  0.275590  0.103870  0.048488  0.055241  0.917394  0.234481   
3     0.858724  0.134243  0.011088  0.000898  0.000000  0.235214  0.010760   
4     0.000000  0.004947  0.005324  0.000000  0.002328  0.038095  0.010919   
5     0.588005  0.043992  0.011968  0.047348  0.003097  1.456227  0.002933   
...        ...       ...       ...       ...       ...       ...       ...   
6036  2.801127  0.783116  0.147585  0.232319  0.071068  1.286187  0.290227   
6037  1.311816  0.108709  0.005989  0.005523  0.005173  0.101903  0.004900   
6038  0.690062  0.017687  0.003195  0.003911  0.001788  0.000000  0.012025   
6039  1.375801  0.283240  0.115842  0.045238  0.064826  0.000000  0.435954   
6040  1.567813  0.051666  0.004411  0.104014  0.000000  0.305785  0.007689   

mID       8         9         10    ...      3943      3944      3945  \
uID                                 ...                                 
1     0.095119  0.000353  0.118706  ...  0.010742  0.000675  0.016204   
2     0.005889  0.111509  0.997125  ...  0.008580  0.001229  0.000066   
3     0.012739  0.052908  0.377859  ...  0.000125  0.000000  0.001965   
4     0.000000  0.000000  0.097795  ...  0.000889  0.000000  0.000441   
5     0.003756  0.000000  0.150823  ...  0.097931  0.000000  0.000106   
...        ...       ...       ...  ...       ...       ...       ...   
6036  0.063616  0.024812  0.468186  ...  0.190765  0.004828  0.007975   
6037  0.000098  0.000000  0.054741  ...  0.005128  0.005982  0.000054   
6038  0.003189  0.000000  0.029432  ...  0.001974  0.000080  0.000548   
6039  0.048027  0.000000  0.122421  ...  0.005431  0.010092  0.007720   
6040  0.007132  0.000000  0.071550  ...  0.094713  0.003164  0.000077   

mID       3946      3947      3948      3949      3950      3951      3952  
uID                                                                         
1     0.007880  0.003810  0.138735  0.049523  0.008876  0.006500  0.089506  
2     0.023668  0.008602  0.020905  0.031919  0.011076  0.008405  0.071477  
3     0.010267  0.000000  0.017614  0.000000  0.000000  0.000000  0.000000  
4     0.000000  0.012629  0.000000  0.000000  0.004538  0.000000  0.000000  
5     0.008791  0.032171  0.108294  0.495394  0.034576  0.036392  0.154063  
...        ...       ...       ...       ...       ...       ...       ...  
6036  0.065516  0.091378  0.227637  0.811697  0.190822  0.148541  0.648291  
6037  0.002267  0.038478  0.018490  0.004392  0.031785  0.001157  0.012081  
6038  0.000493  0.000000  0.014711  0.007754  0.001940  0.002042  0.008285  
6039  0.000000  0.045789  0.051503  0.000000  0.035182  0.000000  0.000000  
6040  0.021432  0.026678  0.102515  0.390906  0.096435  0.084427  0.342959  

[6040 rows x 3664 columns]

In [72]:
def predict_rating(uID, mID):
    try:
        return R_pred.loc[uID, mID]
    except KeyError:
        return 3  # If the user or movie aren't in the train set, return 3

In [73]:
test['pred_rating'] = test.apply(lambda x: predict_rating(x['uID'], x['mID']), axis=1)

Finally, calculate the RMSE

In [74]:
rmse = mean_squared_error(test['rating'], test['pred_rating'], squared=False)
print("RMSE:", rmse)


RMSE: 2.8591626920557047


I obtained a RMSE of: 1.107.

These are the RMSE I obtained in Week 3 assignment.

- Method  |  RMSE
- Baseline Yp = 3
    - 1.2586
- Yp = Mean
    - 1.0353
- Conten based, item-item
    - 1.0128
- Collaborative: Cosine
    - 1.0263
- Collaborative: jaccard Mr >= 3
    - 0.9819
- Collaborative: jaccard Mr >= 1
    - 0.9914
- Collaborative: jaccard Mr
    - 0.9517

## Part 2
Discuss the results and why they did not work well compared to simple baseline or similarity-based methods we’ve done in Module 3. Can you suggest a way(s) to fix it?

The reasons on why the NMF did not that well are:
- The user-movie matrix is large, and NMF is not that optimized for such matrices, so learning latent representation it's more difficult.
- NMF requires n_components, if it's small, the representation is worse, but if it's too large it can overfit.
- Increase max_iterations in NMF.
- It doesn't use similarity as similirarity-based methods, because they use the genre movie information for example, while NMF doesn't.
- Cold start problem: Users with few ratings can be difficult for NMF.
- I fill not rated movies with 0's, which can affect the performance.

Ways to fix:
- Tune the n_components for NMF, in oder to select a balanced n.
- Fill not rated movies with 3's, this can improve.
- Add NMF with global baseline (mean rating) for rare users/movies.

I'll try setting not rated movies with 3's, and see how it performs.

In [75]:
user_movies_matrix = train.pivot(index='uID', columns='mID', values='rating').fillna(3)

In [76]:
nmf = NMF(n_components=20)
user_matrix = nmf.fit_transform(user_movies_matrix)
movies_matrix = nmf.components_

In [77]:
R_pred = np.dot(user_matrix, movies_matrix)
R_pred = pd.DataFrame(R_pred, index=user_movies_matrix.index, columns=user_movies_matrix.columns)

In [78]:
test['pred_rating_3s'] = test.apply(lambda x: predict_rating(x['uID'], x['mID']), axis=1)

In [80]:
rmse = mean_squared_error(test['rating'], test['pred_rating_3s'], squared=False)
print("RMSE fill null with 3:", rmse)

RMSE fill null with 3: 1.0981529239797168


It improves a lot, but it's not yet better than Yp = Mean RMSE (1.0353)
- Next steps it's increasing the n_components for NMF and increase n_interations.